### DISCUSSION TITLE
separate audio for each sentence with corresponding file name

### DISCUSSION THREAD

**Discussion by omega3** [rhasspy/piper#238] [Q&A]
I use piper installed with pipx. I have a spreadsheet with two columns, first column is a sentence number, second column contains sentences. Like this:
01 This is sentence one.
02 This is sentence two.

For each sentence I want to produce separate audio file.

Can you help me and tell me how to use piper to produce audio output files that as a name have number of the sentence, so that I know which file contains which sentence?

It may be exported as csv.

I know I need to use:

```
cat sample.txt | piper -m /home/user/en_GB-southern_english_fema
le-low.onnx --output_file output.wav
```

1. Do I need to have a separate file for each single sentence and loop this command or can I have all sentences in one file?

2. If loop is needed how can I do it?

3.  Do I need some breaks like a second or so, between each query to not overload piper?

**Comment by Sanglyon** [comment]
In the readme, it says:

```
The piper executable can accept JSON input when using the --json-input flag. Each line of input must be a JSON object with text field. For example:

{ "text": "First sentence to speak." }
{ "text": "Second sentence to speak." }
```

and also that  there's an optional parameter: output_file - string 

So you need to format your data in a json file like so: 

```
{ "text":"This is sentence one","output_file":"01.wav"}
{ "text":"This is sentence two","output_file":"02.wav"}

```
and run piper with the parameters:
~~piper -m /home/user/en_GB-southern_english_female-low.onnx **--json-input yourfile.json**~~
edit: see my comment below, the correct way is :

**cat yourfile.jso |** ./piper --model /home/user/en_GB-southern_english_female-low.onnx  **--json-input**

If you have LibreOffice Calc, you can import your csv in a first sheet, 
and in a second sheet:
in the first cell of the first column, the formula  ="{""text"":""" & Sheet1.B1 & """"
in the first cell of the second column, put the formula =" ""output_file"" : """ & Sheet1.A1 & ".wav"" }"
Select A1 to A(number of sentences), Ctrl-D to apply the formula to all cell, repeat for column B,
then export this second sheet as a comma separated csv with no text separator, 
and finally rename the .csv as .json

(You can do the same in Excel, but I don't use it, so I explained with the one I know)

Or write a script to reformat the csv.

**Reply by omega3** [reply]
Could you please share example.csv, i.e. upload it somewhere, like google drive or something like that. I pasted these formulas to LO Calc and I get error 509.

**Reply by Sanglyon** [reply]
It's probably because I put commas after the formulas that weren't part of it, so you may have copy-pasted them by mistake. I've removed them from my explanation above. 
Anyway, here's an exemple  file, but since my locale is french, I don't know if it will automatically translate "Feuille1" to "Sheet1" in english, or your own locale if different. 

[Demo.ods](https://github.com/rhasspy/piper/files/12920706/Demo.ods)

**Reply by omega3** [reply]
I get error. I installed piper with pipx.
```
piper -m /home/user/en_GB-southern_english_female-low.onnx --json-input demo.json
usage: piper [-h] -m MODEL [-c CONFIG] [-f OUTPUT_FILE] [-d OUTPUT_DIR] [--output-raw] [-s SPEAKER]
             [--length-scale LENGTH_SCALE] [--noise-scale NOISE_SCALE] [--noise-w NOISE_W] [--cuda]
             [--sentence-silence SENTENCE_SILENCE] [--data-dir DATA_DIR] [--download-dir DOWNLOAD_DIR] [--update-voices]
             [--debug]
piper: error: unrecognized arguments: --json-input demo.json
```
[demo.json](https://github.com/rhasspy/piper/files/12925724/demo.json)

**Reply by Sanglyon** [reply]
You're right, I had the same error. The argument is in the C++ source, and should appear when calling "piper -h", but it doesn't. 
 I used "pip install piper-tts",  unfortunatly this installed version doesn't seem up-to-date with the git repo. 

I finally got it working by building from source to get the latest version. 
Then after digging in the code, I understood piper needs to be launched this way:
`
cat demo.json | ./piper --model mymodel.onnx --json-input
`

**Reply by omega3** [reply]
Reported: https://github.com/rhasspy/piper/issues/245
